### PR TITLE
Add --haddock-output-dir flag to cabal haddock.

### LINKED ISSUE #8270: Add flag to direct documentation to given location
`--docdir` and `--htmldir` flags do not affect the location of documentation that is generated by `cabal haddock`.

To reproduce the bug I have a minimal project with the following structure:

```
Test
  Test
    Test.hs
  Test.cabal
```

`Test.hs` is an empty module:

```
{-|
Description: Test

Test
-}
module Test.Test () where
```

`Test.cabal`:

```
cabal-version: 3.0
name: Test
version: 0.0.0
library
  build-depends: base
  exposed-modules: Test.Test
```

I would assume that the following flags would affect the location of documentation and put it in `Test/Docs` next to the source folder and the cabal file. But they don't:

```
cabal haddock --docdir Docs
cabal haddock --docdir=Docs
cabal haddock --htmldir Docs
cabal haddock --htmldir=Docs
```

My operating system: Windows 10
cabal: 3.2.0.0
ghc: 8.10.1

### PR DESCRIPTION
**Pull request by LiisiKerik**
This flag gives the user full control over the directory where the documentation is placed. Closes #8270

Tested manually. Documentation for the new flag is displayed with --help. Behavior without the new flag seems the same. The new flag works both in cabal.project and from the command line and behaves as expected. No new unit tests added.

### CODE REVIEW

**Comment by LiisiKerik** [comment]
How do I link this PR to the issue?

**Comment by ffaf1** [comment]
> How do I link this PR to the issue?

Edit your OP and add “closes #xyzx” where `xyzx` is the number of your ticket. Then you will see it in ”Development” in the right bar, both in this PR and in the issue itself.

**Comment by ulysses4ever** [comment]
We certainly need a test.

**Comment by LiisiKerik** [comment]
> We certainly need a test.

Could you please give me some advice and direction for writing a test (where to add the test, how the feature should be tested)? This is my first time using Haskell outside uni or hobby projects. I tried looking at the test suite but found it very confusing.

**Comment by ulysses4ever** [comment]
First, it's readme at https://github.com/haskell/cabal/tree/master/cabal-testsuite (don't read past FAQ though).

Second, try to find somewhat similar test in https://github.com/haskell/cabal/tree/master/cabal-testsuite/PackageTests, copy-paste and edit it... I'd probably look into tests under `*Haddock*` directories. E.g. you could probably go close to https://github.com/haskell/cabal/blob/master/cabal-testsuite/PackageTests/NewHaddock/DisableDoc/cabal.test.hs but with checking output directories and files as in https://github.com/haskell/cabal/blob/master/cabal-testsuite/PackageTests/Haddock/setup.test.hs

**Comment by ulysses4ever** [comment]
@LiisiKerik do you need more directions w.r.t. tests other than what I wrote here https://github.com/haskell/cabal/pull/8788#issuecomment-1441189073? I hope we can bring this over the line soon.

**Comment by LiisiKerik** [comment]
Than you, those references to other test files were very useful. I wrote a couple of tests (which at first seemed to behave reasonably) but then noticed the existence of `--open` flag and started to investigate how it works with the new flag. The following issue came up. When we run `cabal haddock --haddock-odir=Docs --open A` in a folder (let's say `RunFromHere`) which contains folder `A` the docs get placed in `RunFromHere/A/Docs` but the `--open` flag looks for docs in `RunFromHere/Docs` and fails. So now I'm trying to ensure that the docs get placed in the path relative to the original folder, not to the argument folder.

**Comment by LiisiKerik** [comment]
@ulysses4ever @Kleidukos @coot I added a couple of tests and the bug mentioned above is fixed. I'm not sure I picked the best place to make the --haddock-odir path absolute - it would be nice if there was a better place or way to do it.

**Comment by Kleidukos** [comment]
@ulysses4ever Personally I'm inclined to merge it and do some manual QA on it.

File: Cabal/src/Distribution/Simple/Flag.hs [hunk 1]
```diff
@@ -67,6 +67,14 @@ instance Applicative Flag where
   NoFlag   <*> _ = NoFlag
   pure = Flag
 
+instance Foldable Flag where
```
Comment (ulysses4ever) [inline Cabal/src/Distribution/Simple/Flag.hs:70]: These two instances should be derived instead perhaps.

Comment (LiisiKerik) [reply Cabal/src/Distribution/Simple/Flag.hs]: 👍 Done.

**Comment by LiisiKerik** [comment]
@coot @Mikolaj Is there something else I should fix in order to make it ok to merge? @Kleidukos If I understood the comment correctly then you thought it was ok to merge? If you have already looked at my code and decided that it's ok then could you please mark this PR as reviewed?

**Review by Kleidukos — APPROVED** [review]

**Comment by Mikolaj** [comment]
@LiisiKerik: thank you very much the tests. This is a big PR, so we just need to throw a lot pairs of eyes at it. What is your design for backward compatibility here? Are there any corner cases in which people's workflows (e.g, in CI or scripts) stop working after this PR? Unexpected results for workflows that keep working?

**Comment by LiisiKerik** [comment]
@Mikolaj What are the possible backward compatibility issues? I probably don't understand the big picture very well since it's my first pull request in cabal.

> Are there any corner cases in which people's workflows (e.g, in CI or scripts) stop working after this PR? Unexpected results for workflows that keep working?

What would be a good way to find out? I'm only aware that the automatic checks in the server all pass but what other things could I do to find the possible issues?

**Comment by Mikolaj** [comment]
Pardon me, if I'm asking silly questions. You know your code. Did you change any defaults for the other flags, for example? Is the behaviour of cabal with the default setting of the new flag (when the user does not specify it) the same as it was before this PR?

Another angle would be a few manual tests. Move your `~/.cabal` aside, `mkdir ~/.cabal`, and run `cabal update` with a binary of cabal that includes your PR. Then switch to an old cabal, e.g., version 3.4 and try to build a package and some haddocks (you'd need to remove a Nix flag from  `~/.cabal` after it fails parsing it, because we neglected precisely this kind of testing for cabal 3.10).

Then test the other way around, to see how new cabal works with old `~/.cabal` config files.

Perhaps try a the newest haddock and with one from a year ago.

Your PR is substantial (and from I can quickly see, it also changes and improves the codebase), hence the extra burden. I apologies that our CI does not do such tests automatically, but tests that mix up versions of tools are especially tricky to automate and then to triage when they fail.

**Comment by Mikolaj** [comment]
@LiisiKerik: thank you so much.

**Comment by LiisiKerik** [comment]
> The `cabal haddock-project` command has an `--output` flag. I think both commands should use the same name for the flag.

I picked the name by analogy with the other `cabal haddock` flags so that it would look consistent with `haddock-css`, `haddock-hoogle`, `haddock-html` and the other flags that are named according to their plain haddock counterparts. Plain haddock has the `--odir` flag, so `--haddock-odir` (or just `--odir`) would be instantly recognisable for many users.

Most `cabal haddock` flags start with the `--haddock-...` prefix, which I find useful considering how `cabal haddock --help` displays the documentation for all the flags, including those that aren't relevant to `cabal haddock`.

Also, does the `--output` flag for `cabal haddock-project` actually do the exact same thing? Does it place the documentation in the exact directory specified without any additions (for example, if we run the command in `myProject` and specify that `--output=docs` then is the index placed in `myProject/docs/index.html`, outside `dist-newstyle`)? And does it allow relative paths? Sorry for the questions, I haven't read the code of `cabal haddock-project` much and don't understand it well. But if the behavior and purpose of the two flags is different then an identical name would be a source of confusion. If their behavior is indeed the same, then an identical name might be a good idea.

I'm not against changing the name to `--output`, but before changing the name I'd like to ask the other reviewers, too. @ulysses4ever @Kleidukos what do you think?

**Comment by Kleidukos** [comment]
The way I envision it, `haddock-project` being a new addition to the command-line, it is in a "technology preview" phase, and thus we can allow it to diverge a bit in terms of options. However it's not to be taken as standard _yet_. 

Hence, I say that `--haddock-odir`, but be documented also as a tech preview in the manual (and could be subject to change) (Also we need to start tracking those kind of things).

**Comment by coot** [comment]
I didn't say which one should change :), and I am fine if `haddock-project` should actually change it's flag.

> does the --output flag for cabal haddock-project actually do the exact same thing?

It will place the haddocks in a subdirectory of the specified directory, because it handles all components which are present in a given cabal project.

> Does it place the documentation in the exact directory specified without any additions (for example, if we run the command in myProject and specify that --output=docs then is the index placed in myProject/docs/index.html, outside dist-newstyle)?

First part is answered; For the other part, yes it will place everything self contained in the otuput directory which is outside of `dist-newstyle`.

> And does it allow relative paths?

It does.

**Comment by Kleidukos** [comment]
@coot Yeah :) but I don't want to pressure you to change anything about it the flags of `haddock-project` yet. :)

**Comment by LiisiKerik** [comment]
> Hence, I say that we keep `--haddock-odir`, but have it be documented also as a tech preview in the manual (and could be subject to change) (Also we need to start tracking those kind of things).

Where would be a good place to document that it's a tech preview?

**Comment by Kleidukos** [comment]
@LiisiKerik user guide and command-line description :)

**Comment by LiisiKerik** [comment]
@Kleidukos Thank you, I had completely forgotten that the manual also needs to be updated. Now the new flag is documented in cabal-project.rst. And I also added a note that it's a tech preview.

File: doc/cabal-project.rst [hunk 2]
```diff
@@ -1509,6 +1509,15 @@ running ``setup haddock``.
...
     There is no command line variant of this flag.
 
+.. cfg-field:: haddock-odir: path
+               --haddock-odir=PATH
+    :synopsis: Generate haddock documentation into this directory.
+
+    Generate haddock documentation into this directory instead of the default
+    location next to other build products.
+
+    Tech preview, subject to change.
```
Comment (Kleidukos) [inline doc/cabal-project.rst:1519]:
```suggestion
    This flag is provided as a technology preview and is subject to change in the next releases.
```

File: Cabal/src/Distribution/Simple/Setup/Haddock.hs [hunk 3]
```diff
@@ -505,6 +514,11 @@ haddockProjectOptions _showOrParseArgs =
      "location of Haddocks static / auxiliary files"
      haddockProjectLib (\v flags -> flags { haddockProjectLib = v})
      (reqArgFlag "DIR")
+
+    ,option "" ["odir"]
+     "Generate haddock documentation into this directory. (Tech preview, subject to change.)"
```
Comment (Kleidukos) [inline Cabal/src/Distribution/Simple/Setup/Haddock.hs:519]:
```suggestion
     "Generate haddock documentation into this directory. This flag is provided as a technology preview and is subject to change in the next releases."
```

**Review by Kleidukos — CHANGES_REQUESTED** [review]
You got the spirit but here is a correction on the wording :)

**Comment by coot** [comment]
> And I also added a note that it's a tech preview.

I think @Kleidukos meant that `hadock-project` command is a tech preview not the `haddoc-odir` flag, if I am not mistaken.

File: Cabal/src/Distribution/Simple/Setup/Haddock.hs (hunk 3 above)
Comment (coot) [reply Cabal/src/Distribution/Simple/Setup/Haddock.hs:519]: oh, are you sure that something as simple as `--haddock-odir` should be a _tech preview_.  It is consistent with other flags, so I wouldn't bother with it.  Unlike the whole `hadock-project` which deserve to be in that category.

**Comment by fgaz** [comment]
I'd at least change it to something more verbose, like one of --haddock-output{-dir,dir,}

I think "odir" is too unclear

**Comment by mouse07410** [comment]
I'm ambivalent, but `--haddock-odir` looks clear enough to me, especially when it's documented.

File: Cabal/src/Distribution/Simple/Setup/Haddock.hs (hunk 3 above)
Comment (Kleidukos) [reply Cabal/src/Distribution/Simple/Setup/Haddock.hs:519]: Because you never know in advance what may happen, and it costs us nothing.

**Comment by Kleidukos** [comment]
Frankly I'd like to drop the "odir" term because indeed it's not great for discoverability (people will reasonably look up "output"). Having a fully-worded `output` in the flag name is indeed very attractive.

**Comment by LiisiKerik** [comment]
I changed the flag to `--haddock-odir`.

**Comment by LiisiKerik** [comment]
@coot I'm not sure why it says "requested review from [Kleidukos](https://github.com/Kleidukos) and removed request for [coot](https://github.com/coot) [1 minute ago](https://github.com/haskell/cabal/pull/8788#event-8863774047)" I didn't try to remove you from the reviewer list. I just clicked something to re-request a review after making the suggested changes...

**Review by Kleidukos — CHANGES_REQUESTED** [review]
We should use `outputDir` more generally in the code base instead of `odir`. :)

**Comment by mouse07410** [comment]
It would **definitely** be clearer for the user.

**Comment by LiisiKerik** [comment]
I changed the name to `--haddock-output-dir`.

File: Cabal/src/Distribution/Simple/Flag.hs [hunk 4]
```diff
@@ -67,6 +69,14 @@ instance Applicative Flag where
   NoFlag   <*> _ = NoFlag
   pure = Flag
 
+{-instance Foldable Flag where
+  foldMap _ NoFlag = mempty
+  foldMap f (Flag x) = f x
+
+instance Traversable Flag where
+  sequenceA NoFlag = pure NoFlag
+  sequenceA (Flag x) = Flag <$> x-}
```
Comment (Kleidukos) [inline Cabal/src/Distribution/Simple/Flag.hs:78]: I guess you can remove that code block now?

**Review by Kleidukos — APPROVED** [review]
Perfect. You can change the name of the PR to reflect the new name of the flag but otherwise it's all good.

**Comment by LiisiKerik** [comment]
@coot @ulysses4ever Is there something else I should fix in order to make it ok to merge?

**Comment by ulysses4ever** [comment]
One more thing: why the Show instances? Are they really needed for anything? Bear in mind that `deriving` is not free: we pay for it with the time on every compilation.

**Comment by ulysses4ever** [comment]
> Personally I'm inclined to merge it and do some manual QA on it.

Just for the record: I applaud the manual QA effort but I’d be frustrated if we’re gonna use it as an excuse for not adding tests. I’m very glad that the PR author added some.

File: doc/cabal-project.rst [hunk 5]
```diff
@@ -1509,6 +1509,16 @@ running ``setup haddock``.
 
     There is no command line variant of this flag.
 
+.. cfg-field:: haddock-output-dir: path
+               --haddock-output-dir=PATH
+    :synopsis: Generate haddock documentation into this directory.
```
[the remark this answers is not in the document]
Comment (LiisiKerik) [reply doc/cabal-project.rst:1514]: No, this is not true. Relative paths are allowed. I think that not allowing relative paths would greatly reduce usability and convenience for the end user.

**Comment by LiisiKerik** [comment]
> One more thing: why the Show instances? Are they really needed for anything? Bear in mind that `deriving` is not free: we pay for it with the time on every compilation.

Those were an accidental leftover from testing. I removed those now.

File: cabal-install/src/Distribution/Client/ProjectConfig/Types.hs [hunk 6]
```diff
@@ -291,6 +291,7 @@ data PackageConfig
        packageConfigHaddockIndex        :: Flag PathTemplate, --TODO: [required eventually] use this
        packageConfigHaddockBaseUrl      :: Flag String, --TODO: [required eventually] use this
        packageConfigHaddockLib          :: Flag String, --TODO: [required eventually] use this
+       packageConfigHaddockOutputDir    :: Flag String, --TODO: [required eventually] use this
```
[the remark this answers is not in the document]
Comment (LiisiKerik) [reply cabal-install/src/Distribution/Client/ProjectConfig/Types.hs:294]: Thank you, that does look more informative. Done.

File: changelog.d/issue-8270 [hunk 7]
```diff
@@ -0,0 +1,4 @@
+synopsis: Add --haddock-output-dir flag to cabal haddock.
+packages: Cabal cabal-install
+prs: #8788
```
[the remark this answers is not in the document]
Comment (LiisiKerik) [reply changelog.d/issue-8270:3]: Made the changes, other than the absolute path part which isn't correct. (Although I'm not sure the description is good. I'm open to suggestions about improving the description.)

**Review by ulysses4ever — APPROVED** [review]
Thank you!

One last comment: the changelog file supports Markdown markup. I'd add backticks to commands like `cabal haddock` and the flag name. That way our release notes will look real slick.

**Comment by LiisiKerik** [comment]
> Thank you!
> 
> One last comment: the changelog file supports Markdown markup. I'd add backticks to commands like `cabal haddock` and the flag name. That way our release notes will look real slick.

Does the synopsis field also support markdown? Or only the description?

**Comment by ulysses4ever** [comment]
@LiisiKerik both. We should document it of course...

**Comment by LiisiKerik** [comment]
Since I'm new to the process, I don't know what to do next now that the changes have been marked as approved. Do I (or someone else?) need to change tags on the pull request? Do merges happen continuously or only before releases?

**Comment by ffaf1** [comment]
Add a `squash+merge` tag, then wait patiently for two days.

**Comment by ulysses4ever** [comment]
There's just one commit, so the merge-me label should suffice.

**Comment by LiisiKerik** [comment]
@ffaf1 @ulysses4ever Thank you. But, it seems that I don't have the right to add labels. I don't see the gear icon next to labels.

**Comment by Kleidukos** [comment]
I'll do it, thank you for your patience @LiisiKerik :)

**Comment by LiisiKerik** [comment]
> This pull request cannot be embarked for merge
> The merge queue pull request can't be updated
> Details:
> 
> Unable to update: user LiisiKerik is unknown.
> 
> Please make sure LiisiKerik has logged in Mergify dashboard.

**Comment by ulysses4ever** [comment]
@mergify rebase

**Comment by ulysses4ever** [comment]
@LiisiKerik sorry, it's just a confusing error message: i don't believe you can do anything about it. Let's see if my command helps.

**Comment by ulysses4ever** [comment]
And it's done! @LiisiKerik thanks a lot for the great contribution and patience along the way!
